### PR TITLE
Edge - JSON converter for proto 

### PR DESCRIPTION
**Pull request by AndriiLandiak**
## Pull Request description

**Issue**:
Currently, maintaining backward compatibility with Edge when new fields are added to entities is cumbersome. It involves updating proto configurations and extensive testing after any Tb entity updates.

**Solution**:
To simplify this process, this PR uses serialization and deserialization for sending entities. By using Object Mapper serialization and setting FAIL_ON_UNKNOWN_PROPERTIES to **false** during deserialization, we can seamlessly handle different entity versions between our Edge and TB.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: common/util/src/main/java/org/thingsboard/common/util/JacksonUtil.java [hunk 1]
```diff
@@ -146,6 +150,15 @@ public static String toPrettyString(Object o) {
         }
     }
 
+    public static <T> T fromEdgeString(String string, Class<T> clazz) {
```
Comment (volodymyr-babak) [inline common/util/src/main/java/org/thingsboard/common/util/JacksonUtil.java:153]: please rename method to **fromStringIgnoreUnknownFields** or similar

Comment (volodymyr-babak) [reply common/util/src/main/java/org/thingsboard/common/util/JacksonUtil.java:153]: the same applies for EDGE_OBJECT_MAPPER

File: common/util/src/test/java/org/thingsboard/common/util/JacksonUtilTest.java [hunk 2]
```diff
@@ -32,4 +37,23 @@ public void allow_unquoted_field_mapper_test() {
         Assert.assertThrows(IllegalArgumentException.class, () -> JacksonUtil.toJsonNode(data)); // syntax exception due to missing quotes in the field name!
     }
 
-}
\ No newline at end of file
+    @Test
+    public void fail_on_unknown_properties_edge_mapper_test() throws JsonProcessingException {
```
Comment (volodymyr-babak) [inline common/util/src/test/java/org/thingsboard/common/util/JacksonUtilTest.java:41]: naming convention - please rename it according to default naming convention.
and please do the same for the test above allow_unquoted_field_mapper_test

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/alarm/BaseAlarmProcessor.java [hunk 3]
```diff
@@ -112,6 +108,27 @@ public ListenableFuture<Void> processAlarmMsg(TenantId tenantId, AlarmUpdateMsg
...
+        alarm.setStartTs(alarmUpdateMsg.getStartTs());
+        AlarmStatus alarmStatus = AlarmStatus.valueOf(alarmUpdateMsg.getStatus());
+        alarm.setClearTs(alarmUpdateMsg.getClearTs());
+        alarm.setPropagate(alarmUpdateMsg.getPropagate());
+        alarm.setCleared(alarmStatus.isCleared());
+        alarm.setAcknowledged(alarmStatus.isAck());
+        alarm.setAckTs(alarmUpdateMsg.getAckTs());
+        alarm.setEndTs(alarmUpdateMsg.getEndTs());
+        try {
+            alarm.setDetails(JacksonUtil.OBJECT_MAPPER.readTree(alarmUpdateMsg.getDetails()));
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/alarm/BaseAlarmProcessor.java:125]: JacksonUtil.OBJECT_MAPPER.readTree -> JacksonUtil.toJsonNode

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/asset/AssetProfileEdgeProcessor.java [hunk 4]
```diff
@@ -127,19 +129,24 @@ public DownlinkMsg convertAssetProfileEventToDownlink(EdgeEvent edgeEvent, EdgeI
...
 
     @Override
-    protected void setDefaultEdgeRuleChainId(TenantId tenantId,AssetProfile assetProfile, AssetProfileUpdateMsg assetProfileUpdateMsg) {
-        UUID defaultEdgeRuleChainUUID = safeGetUUID(assetProfileUpdateMsg.getDefaultRuleChainIdMSB(), assetProfileUpdateMsg.getDefaultRuleChainIdLSB());
+    protected void setDefaultEdgeRuleChainId(AssetProfile assetProfile, RuleChainId ruleChainId, AssetProfileUpdateMsg assetProfileUpdateMsg, boolean isEdgeVersionDeprecated) {
+        UUID defaultEdgeRuleChainUUID = isEdgeVersionDeprecated
+                ? safeGetUUID(assetProfileUpdateMsg.getDefaultRuleChainIdMSB(), assetProfileUpdateMsg.getDefaultRuleChainIdLSB())
+                : assetProfile.getDefaultRuleChainId() != null ? assetProfile.getDefaultRuleChainId().getId() : null;
         assetProfile.setDefaultEdgeRuleChainId(defaultEdgeRuleChainUUID != null ? new RuleChainId(defaultEdgeRuleChainUUID) : null);
+        assetProfile.setDefaultRuleChainId(ruleChainId);
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/asset/AssetProfileEdgeProcessor.java:142]: I believe this should be moved to 
**protected void setDefaultRuleChainId(TenantId tenantId, AssetProfile assetProfile) {**
method?

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/dashboard/BaseDashboardProcessor.java [hunk 5]
```diff
@@ -21,57 +21,73 @@
...
+            assignedCustomers = filterNonExistingCustomers(tenantId, assignedCustomers);
+            dashboard.setAssignedCustomers(assignedCustomers);
+        }
         Dashboard savedDashboard = dashboardService.saveDashboard(dashboard, false);
         if (assignedCustomers != null && !assignedCustomers.isEmpty()) {
-            for (ShortCustomerInfo assignedCustomer : assignedCustomers) {
-                if (assignedCustomer.getCustomerId().equals(customerId)) {
-                    dashboardService.assignDashboardToCustomer(tenantId, dashboardId, assignedCustomer.getCustomerId());
-                }
+            for (ShortCustomerInfo assignedCustomer : savedDashboard.getAssignedCustomers()) {
```
[the remark this answers is not in the document]
Comment (volodymyr-babak) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/dashboard/BaseDashboardProcessor.java:64]: Use case: Dashboard has 10 assigned customers.
You are updating dashboard on the edge, where only 0 or 1 customers are present.
How many assignDashboardToCustomer calls are going to be in this case?

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/device/DeviceProfileEdgeProcessor.java [hunk 6]
```diff
@@ -127,19 +129,24 @@ public DownlinkMsg convertDeviceProfileEventToDownlink(EdgeEvent edgeEvent, Edge
...
 
     @Override
-    protected void setDefaultEdgeRuleChainId(TenantId tenantId, DeviceProfile deviceProfile, DeviceProfileUpdateMsg deviceProfileUpdateMsg) {
-        UUID defaultEdgeRuleChainUUID = safeGetUUID(deviceProfileUpdateMsg.getDefaultRuleChainIdMSB(), deviceProfileUpdateMsg.getDefaultRuleChainIdLSB());
+    protected void setDefaultEdgeRuleChainId(DeviceProfile deviceProfile, RuleChainId ruleChainId, DeviceProfileUpdateMsg deviceProfileUpdateMsg, boolean isEdgeVersionDeprecated) {
+        UUID defaultEdgeRuleChainUUID = isEdgeVersionDeprecated
+                ? safeGetUUID(deviceProfileUpdateMsg.getDefaultRuleChainIdMSB(), deviceProfileUpdateMsg.getDefaultRuleChainIdLSB())
+                : deviceProfile.getDefaultRuleChainId() != null ? deviceProfile.getDefaultRuleChainId().getId() : null;
         deviceProfile.setDefaultEdgeRuleChainId(defaultEdgeRuleChainUUID != null ? new RuleChainId(defaultEdgeRuleChainUUID) : null);
+        deviceProfile.setDefaultRuleChainId(ruleChainId);
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/device/DeviceProfileEdgeProcessor.java:142]: probably it should be moved to:
**protected void setDefaultRuleChainId(TenantId tenantId, DeviceProfile deviceProfile) {**
?

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/alarm/BaseAlarmProcessor.java [hunk 7]
```diff
@@ -37,19 +38,28 @@
...
         AlarmId alarmId = new AlarmId(new UUID(alarmUpdateMsg.getIdMSB(), alarmUpdateMsg.getIdLSB()));
+        boolean isEdgeProtoDeprecated = EdgeVersionUtils.isEdgeProtoDeprecated(edgeVersion);
+        Alarm alarm = isEdgeProtoDeprecated ? createDeprecatedAlarm(tenantId, alarmUpdateMsg)
+                : JacksonUtil.fromEdgeString(alarmUpdateMsg.getEntity(), Alarm.class);
+        if (alarm == null) {
+            throw new RuntimeException("[{" + tenantId + "}] alarmUpdateMsg {" + alarmUpdateMsg + "} cannot be converted to alarm");
+        }
+        EntityType entityType = isEdgeProtoDeprecated ? EntityType.valueOf(alarmUpdateMsg.getOriginatorType())
+                : alarm.getOriginator().getEntityType();
+        EntityId originatorId = getAlarmOriginator(tenantId, alarmUpdateMsg.getOriginatorName(), entityType);
```
[the remark this answers is not in the document]
Comment (AndriiLandiak) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/alarm/BaseAlarmProcessor.java:62]: I got your point. We could remove this method from a new approach with JSON, but leave it for backward compatibility.

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/dashboard/BaseDashboardProcessor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndriiLandiak) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/dashboard/BaseDashboardProcessor.java:64]: _How many assignDashboardToCustomer calls are going to be in this case?_
**Asnwer: 0 or 1 :)** 

image
image

**Pull request closed without merging**